[wip] E35

awskii opened the pull request.

AskAlexSharov commented on state/aggregator_v3.go:
```diff
@@ -1045,29 +1325,44 @@ func (a *AggregatorV3) mergeFiles(ctx context.Context, files SelectedStaticFiles
+		predicates.Add(1)
+		g.Go(func() (err error) {
+			mf.storage, mf.storageIdx, mf.storageHist, err = a.storage.mergeFiles(ctx, files.storage, files.storageIdx, files.storageHist, r.storage, workers)
+			predicates.Done()
```
always call `wg.Done` in `defer` to avoid deadlocks on `panic`

AskAlexSharov commented on state/domain.go:
```diff
@@ -33,6 +33,7 @@ import (
 
 	"github.com/RoaringBitmap/roaring/roaring64"
 	btree2 "github.com/tidwall/btree"
+	atomic2 "go.uber.org/atomic"
```
we dropped support of go1.18 in last release. now can use `sync/atomic.Int32`

AskAlexSharov commented on state/domain_mem.go:
```diff
@@ -0,0 +1,443 @@
+}
+
+func (l *KVList) Reset() {
+	//l.Keys = l.Keys[:0]
```
I prefer as the first step: get much green tests. And ten introduce prealloc/reset optimizations.

AskAlexSharov commented on state/domain_mem.go:
```diff
@@ -0,0 +1,443 @@
+	*Domain
+
+	etl    *etl.Collector
+	mu     sync.RWMutex
```
do we need shared mutex between domains?

AskAlexSharov commented on state/domain_mem.go:
```diff
@@ -0,0 +1,443 @@
+	//}
+	return func(k []byte, value []byte, _ etl.CurrentTableReader, next etl.LoadNextFunc) error {
+		// if its ordered we could put to history each key excluding last one
+		tx := binary.BigEndian.Uint64(k[len(k)-8:])
```
`tx` variable name is reserved for Database Transaction

awskii committed:
save

awskii committed:
Merge branch 'e35' of github.com:ledgerwatch/erigon-lib into e35

awskii committed:
save

awskii committed:
restore locality

AskAlexSharov committed:
Merge branch 'e35' into e35_ex

AskAlexSharov committed:
Merge branch 'e35_ex' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35' into e35_ex

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35_ex' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35_ex' into e35

AskAlexSharov committed:
save

awskii committed:
save

awskii committed:
save

AskAlexSharov committed:
Merge branch 'e35' into e35_ex

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35_ex' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
save

awskii committed:
save

awskii committed:
working save

awskii committed:
working save

awskii committed:
save

awskii committed:
revert

awskii committed:
save fix iter

awskii committed:
save

AskAlexSharov committed:
return nil instead of ErrBtIndexLookupBounds

awskii committed:
save

awskii committed:
save nolocality cold

awskii committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
save

awskii committed:
save

awskii committed:
save

awskii committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
testcase: locality index to kv file

AskAlexSharov committed:
disable ii compression

AskAlexSharov committed:
disable ii compression

awskii committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
step towards 1 seed

awskii committed:
save

awskii committed:
Merge branch 'e35' of github.com:ledgerwatch/erigon-lib into e35

awskii committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
fix

awskii committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'test_non_nil_next' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35_same_seed3' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
save

awskii committed:
save

awskii committed:
save

AskAlexSharov committed:
try commit

AskAlexSharov committed:
save

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
TestSharedDomain_Unwind: deadlock on panic fix

AskAlexSharov committed:
save

awskii committed:
fix

awskii committed:
fix

awskii committed:
fix

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
fix

awskii committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
merge devel

AskAlexSharov committed:
Merge branch 'main' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'e35' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

awskii committed:
save

awskii committed:
Merge branch 'e35' of github.com:ledgerwatch/erigon-lib into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
domain: avoid extra bt.Has()

AskAlexSharov committed:
domain_shared: replace bt by map.
history: pre-alloc stateless ArchiveGetter
ef: add Seek method for future optimizations

AskAlexSharov committed:
domain: use pre-allocated cursors

AskAlexSharov committed:
simplify MakeContext

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
mend

crypto zero-copy

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
inlinable select

AskAlexSharov committed:
inlinable select

AskAlexSharov committed:
save

AskAlexSharov committed:
Revert "inlinable select"

This reverts commit e6e17a06689c73f6f1d2dc771fe350507a89a2af.

AskAlexSharov committed:
Revert "save"

This reverts commit 27eb8714aab10513ecf23fa818213b5e1a5104f7.

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
optimize MakeContext

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'downloader_faster_scan' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 't3' into e35_sharded_bloom_ef

AskAlexSharov committed:
Merge branch 'e35_sharded_bloom_ef' into e35

AskAlexSharov committed:
save

AskAlexSharov committed:
fix linter error

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
save

AskAlexSharov committed:
Merge branch 'main' into e35_sharded_bloom_ef

AskAlexSharov committed:
Merge branch 'e35_sharded_bloom_ef' into e35

AskAlexSharov committed:
save

awskii committed:
save
